### ISSUE TITLE
Why loss is multiplied by world size

### ISSUE THREAD

**Issue by fzyzcjy** [ChenmienTan/RL2#11]
Hi thanks for the lib! Maybe a very silly question: It seems that the loss value for backprop is world_size times larger than the loss value for metrics. I checked https://docs.pytorch.org/tutorials/intermediate/FSDP_tutorial.html and does not find this multiplication.

![Image](https://github.com/user-attachments/assets/b1e18607-b25a-4412-832b-0738e9d7e823)

**Comment by ChenmienTan** [comment]
This is because FSDP will average gradient across devices.

In the standard setting, each device processes identical number, i.e., `batch_size_per_device`, of samples. In this case, you can simply back-propagate the average loss on each device. Since FSDP will average gradient across devices, this is equivalent to back-propagate the average loss of all samples.

However, we use balanced sequence packing for higher throughput so different devices may process different number of samples. To compute the average loss of all samples, on each device, we firstly sum the losses of samples on the device and divide the sum by the total number of samples, which should be **summed** across devices. If we do not multiply the value by world size before back-propagation, since FSDP will average gradient across devices, it is equivalent to averaging the value rather than summing it.
